### ISSUE TITLE
Spring Boot 3 Native Image support

### ISSUE THREAD

**Issue by justin-tay** [yahoo/elide#2934]
Would be good if Elide can incorporate support to be compiled as a native image in Spring Boot 3

I noticed the following issues when trying to build a native image

* `ElideConfigProperties` is missing `NestedConfigurationProperty` annotations. Previously it just affected some non essential meta data but during AOT processing this means the reflection meta data configuration is not generated for the nested configuration classes which then silently fail to bind the config properties.
* The `FileLoader` needs changes as Graal will use resources with the `resource` protocol.
* The cache used by the `DefaultClassScanner` needs to be modified to be statically initialized and configured to be loaded at build time `--initialize-at-build-time` along with ClassGraph itself
* A `org.graalvm.nativeimage.hosted.Feature` can be created to register the reflection metadata of classes scanned by ClassGraph
* `@RefreshScope` has issues as a native image so there needs to be a way to make it optional. See https://docs.spring.io/spring-cloud-config/docs/current/reference/html/#_aot_and_native_image_support_2.
* The libraries can incorporate `native-image.properties` to register the approprate metadata of Elide and some of its dependencies

**Comment by aklish** [comment]
Hi Justin,

This feels similar to work I did to make Elide work with Quarkus and
Graal.  Definitely worth doing.  I think Elide startup was up under 3ms on
my mac when running under native.   Looks like you've done some exploration
here.  Do you have anything on a branch that we could work off of?  Thanks.

Aaron

**Comment by justin-tay** [comment]
Am working on the following branches

elide
* https://github.com/justin-tay/elide/tree/native

elide-spring-boot-example
* https://github.com/justin-tay/elide-spring-boot-example/tree/native_elide-7.x

In my manual tests I think it should generally be working. I haven't been able to get rest-assured working in a native test due to groovy.

**Comment by aklish** [comment]
This looks great.  I'm building your branch now.  A few questions on the branch:

1.  Is there a problem with GraalVM and System.lineSeparator();?  I've noticed those lines were replaced.
2. How did you build the reflect-config.json and resource-config.json files?  Was that automated somehow or did you have to hunt down errors for the reflection calls?

**Comment by aklish** [comment]
Also - would you be willing to make this a PR against main Elide?

**Comment by justin-tay** [comment]
1. No there's no issue with GraalVM and System.lineSeparator. It's just that actually quite a number of the pre-existing tests are failing when being run in Windows due to the line separator differences and I needed to get the library to build cleanly in case I caused any regressions. The pre-existing tests assume `\n` while on Windows it is `\r\n` and are hence failing. There were also some other minor issues on Windows which I fixed. I ran the tests again on a Mac M1 so I don't think this breaks anything.

2. It's largely manual but first I run the application using the native tracing agent and exercise the application first to let it first generate all the metadata. However it typically generates too much metadata, and using some of that metadata as is can also lead to the native build failing. I then run it natively and then see where it runs into errors and selectively copy the metadata over and repeating the process until it all gets resolved.
